src/core/lib/resource_quota/memory_quota.h

```diff
@@ -0,0 +1,322 @@
+    MaybeRegisterReclaimer();
+  }
+
+  // Post a reclaimer for some reclamation pass.
+  void PostReclaimer(ReclamationPass pass,
+                     std::function<void(ReclamationSweep)>);
+
+  // Allocate a new object of type T, with constructor arguments.
+  // The returned type is wrapped, and upon destruction the reserved memory
+  // will be released to the allocator automatically. As such, T must have a
+  // virtual destructor so we can insert the necessary hook.
+  template <typename T, typename... Args>
+  absl::enable_if_t<std::has_virtual_destructor<T>::value, T*> New(
+      Args&&... args) ABSL_LOCKS_EXCLUDED(memory_quota_mu_) {
+    // Wrap T such that when it's destroyed, we can release memory back to the
+    // allocator.
+    class Wrapper final : public T {
+     public:
+      explicit Wrapper(RefCountedPtr<MemoryAllocator> allocator, Args&&... args)
+          : T(std::forward<Args>(args)...), allocator_(std::move(allocator)) {}
+      ~Wrapper() override { allocator_->Release(sizeof(*this)); }
+
+     private:
+      const RefCountedPtr<MemoryAllocator> allocator_;
+    };
+    Reserve(sizeof(Wrapper));
+    return new Wrapper(Ref(), std::forward<Args>(args)...);
+  }
+
+  // Construct a unique ptr immediately.
+  template <typename T, typename... Args>
+  std::unique_ptr<T> MakeUnique(Args&&... args)
+      ABSL_LOCKS_EXCLUDED(memory_quota_mu_) {
+    return std::unique_ptr<T>(New<T>(std::forward<Args>(args)...));
+  }
+
+ private:
+  // Primitive reservation function.
+  absl::optional<size_t> TryReserve(MemoryRequest request) GRPC_MUST_USE_RESULT;
+  // Replenish bytes from the quota, without blocking, possibly entering
+  // overcommit.
+  void Replenish() ABSL_LOCKS_EXCLUDED(memory_quota_mu_);
+  // If we have not already, register a reclamation function against the quota
+  // to sweep any free memory back to that quota.
+  void MaybeRegisterReclaimer() ABSL_LOCKS_EXCLUDED(memory_quota_mu_);
+  void MaybeRegisterReclaimerLocked()
+      ABSL_EXCLUSIVE_LOCKS_REQUIRED(memory_quota_mu_);
+
+  // Amount of memory this allocator has cached for its own use: to avoid quota
+  // contention, each MemoryAllocator can keep some memory in addition to what
+  // it is immediately using, and the quota can pull it back under memory
+  // pressure.
+  std::atomic<size_t> free_bytes_{0};
+  // Mutex guarding the backing resource quota.
+  Mutex memory_quota_mu_;
+  // Backing resource quota.
+  RefCountedPtr<MemoryQuota> memory_quota_ ABSL_GUARDED_BY(memory_quota_mu_);
+  // Amount of memory taken from the quota by this allocator.
+  size_t taken_bytes_ ABSL_GUARDED_BY(memory_quota_mu_) = 0;
+  // Indices into the various reclaimer queues, used so that we can cancel
+  // reclamation should we shutdown or get rebound.
+  ReclaimerQueue::Index
+      reclamation_indices_[kNumReclamationPasses] ABSL_GUARDED_BY(
+          memory_quota_mu_) = {ReclaimerQueue::kInvalidIndex};
+};
+
+class AtomicBarrier {
```

Comment:
Some light class documentation would be appreciated.

Reply:
Yeah... see the notes on this PR... still trying to figure out what becomes of this.